### PR TITLE
porting for CH32V203

### PR DESCRIPTION
**Pull request by TianpeiLee**
Hello, I have adapted on the CH32V203, and added the MRS project folder of WCH, and I verified it on the hardware provided by airbot, it works very well, thank you very much for the support of airbot.

I've changed the definition of a couple of variables by adding the keyword "volatile" in front of them to increase their compatibility under gcc, and this will not affect compilation under armcc

### CODE REVIEW

**Comment by tridge** [comment]
I've been looking at adding support for this MCU in AM32-bootloader:
https://github.com/tridge/AM32-bootloader
are you on the AM32 discord? If so please ping me and we could create a thread under development
I have a couple of CH32V203 dev boards, but need some help with getting openocd working, setup of toolchain etc

**Comment by tridge** [comment]
here is a rebase of this PR on top of current main with some fixes:
https://github.com/tridge/AM32/tree/pr-ch32v203-wip

**Comment by tridge** [comment]
I've now got the bootloader working. This AM32 tree:
https://github.com/tridge/AM32/tree/pr-ch32v203-wip
now works with this bootloader:
https://github.com/am32-firmware/AM32-bootloader/pull/4

File: Src/main.c [hunk 1]
```diff
@@ -1680,6 +1683,10 @@ void runBrushedLoop()
 
 int main(void)
 {
+#ifdef DEBUG_SDI
```
Comment (tridge) [inline Src/main.c:1686]: we should remove this extra debug code before merge

File: Src/main.c [hunk 2]
```diff
@@ -1832,6 +1839,12 @@ int main(void)
         }
 #endif
 
+#ifdef MCU_CH32V203
```
Comment (tridge) [inline Src/main.c:1842]: we should add a NEED_INPUT_READY define for F031 and V203, instead of copying the code

File: Src/main.c [hunk 3]
```diff
@@ -1852,6 +1865,7 @@ int main(void)
                 for (int i = 0; i < 64; i++) {
                     dma_buffer[i] = 0;
                 }
+                PRINT("reset1:%d\r\n",signaltimeout);
```
Comment (tridge) [inline Src/main.c:1868]: debug should be removed

File: Src/main.c [hunk 4]
```diff
@@ -1865,6 +1879,7 @@ int main(void)
                 for (int i = 0; i < 64; i++) {
                     dma_buffer[i] = 0;
                 }
+                PRINT("reset2:%d\r\n",signaltimeout);
```
Comment (tridge) [inline Src/main.c:1882]: remove debug

File: Src/main.c [hunk 5]
```diff
@@ -1952,15 +1967,37 @@ int main(void)
         }
 
 #ifndef MCU_F031
+#ifdef WCH
```
Comment (tridge) [inline Src/main.c:1970]: commented out code should be removed

File: Src/main.c [hunk 6]
```diff
@@ -1952,15 +1967,37 @@ int main(void)
...
         if (dshot_telemetry && (commutation_interval > DSHOT_PRIORITY_THRESHOLD)) {
-            NVIC_SetPriority(IC_DMA_IRQ_NAME, 0);
-            NVIC_SetPriority(COM_TIMER_IRQ, 1);
-            NVIC_SetPriority(COMPARATOR_IRQ, 1);
-        } else {
-            NVIC_SetPriority(IC_DMA_IRQ_NAME, 1);
-            NVIC_SetPriority(COM_TIMER_IRQ, 0);
-            NVIC_SetPriority(COMPARATOR_IRQ, 0);
-        }
+             NVIC_SetPriority(IC_DMA_IRQ_NAME, 0);
```
Comment (tridge) [inline Src/main.c:1991]: the whitespace changes here should be reverted

File: Src/main.c [hunk 7]
```diff
@@ -1993,6 +2030,11 @@ int main(void)
             adc_ordinary_software_trigger_enable(ADC1, TRUE);
             //  converted_degrees = (4000 - ADC_raw_temp) / 20;
             converted_degrees = getConvertedDegrees(ADC_raw_temp);
+#endif
+#ifdef WCH
+//            ADC_DMA_Callback( ); //�ж��б�����
```
Comment (tridge) [inline Src/main.c:2035]: comments in english please

File: Src/signal.c [hunk 8]
```diff
@@ -160,7 +160,13 @@ void transfercomplete()
                     buffersize = 2;
                     computeServoInput();
                 }
+#ifdef MCU_CH32V203
```
Comment (tridge) [inline Src/signal.c:163]: why not make these changes part of receiveDShotDma() ?

File: Src/sounds.c [hunk 9]
```diff
@@ -97,7 +97,12 @@ void playStartupTune()
     __disable_irq();
 
     uint8_t value = *(uint8_t*)(EEPROM_START_ADD + 48);
-    if (value != 0xFF) {
+#ifdef WCH
```
Comment (tridge) [inline Src/sounds.c:100]: why is this change needed?

**Review by tridge — CHANGES_REQUESTED** [review]

File: Src/sounds.c (hunk 9 above)
Comment (TianpeiLee) [reply Src/sounds.c:100]: Because after it is  erased, its default value is 0xe339e339

File: Src/main.c (hunk 2 above)
Comment (TianpeiLee) [reply Src/main.c:1842]: Yes, NEED_INPUT_READY would be bettter,  it was distinguished by name before.

File: Src/sounds.c (hunk 9 above)
Comment (tridge) [reply Src/sounds.c:100]: ok, better to have a define ERASED_FLASH_BYTE, we could default it to 0xff at the top of sounds.c, but use ifndef to allow a Mcu port to override it

Comment (TianpeiLee) [reply Src/sounds.c:100]: yeah  would be better

File: Src/signal.c (hunk 8 above)
Comment (TianpeiLee) [reply Src/signal.c:163]: It can indeed be moved to receiveDShotDma()

File: Src/main.c (hunk 1 above)
Comment (TianpeiLee) [reply Src/main.c:1686]: yeah,  all debug related code can be deleted

File: Src/main.c [hunk 10]
```diff
@@ -224,7 +224,7 @@ an settings option)
 #endif
 
 #define VERSION_MAJOR 2
-#define VERSION_MINOR 8
+#define VERSION_MINOR 12
```
Comment (tridge) [inline Src/main.c:227]: the PR needs to be rebased on current main branch, these defines no longer exist

**Review by tridge — CHANGES_REQUESTED** [review]
the PR is looking pretty good now, but needs a rebase on current main branch

File: Src/functions.c [hunk 11]
```diff
@@ -141,3 +141,31 @@ void gpio_mode_QUICK(gpio_type* gpio_periph, uint32_t mode,
 }
 #endif
 #endif
+
+
+#ifdef WCH
+
+void delayMicros(uint32_t micros)
+{
```
Comment (tridge) [inline Src/functions.c:149]: i'm not a fan of the busy loop method, depends too much on the clock selected. Works for current v203 but will fail on other variants. Don't we have a timer we can use?

**Comment by tridge** [comment]
@TianpeiLee I have rebased your code and fixed some issues with it in this branch:
https://github.com/tridge/AM32/tree/pr-ch32v203-wip

File: Src/functions.c (hunk 11 above)
Comment (TianpeiLee) [reply Src/functions.c:149]: All timers have been used

Comment (TianpeiLee) [reply Src/functions.c:149]: Originally, this delay should use the core timer, so that it can be used as long as it is a chip with the same core. However, the V203 has a total of 4 peripheral timers.   

1 for scheduled task, 2 for commutation tasks, 1 for dshot signal detection task, and 1 for PWM output task，so.....

**Comment by tridge** [comment]
the commit "v203_pr" removed thousands of files, removing support for other MCUs, and removing the CI tests
please do a hard reset against my branch 
https://github.com/tridge/AM32/tree/pr-ch32v203-wip

**Comment by tridge** [comment]
@TianpeiLee I also just noticed this PR is missing the handling of the driving_brake_strength in Src/main.c
you need to search for '#ifdef STMICRO' and make sure ch32 is handled in all the same places

**Comment by TianpeiLee** [comment]
@tridge Hi,How about resubmitting the pr directly from your branch? Not only is my modification and update very slow every time, but it also fails due to network reasons.

**Comment by tridge** [comment]
@TianpeiLee new PR opened https://github.com/am32-firmware/AM32/pull/89

**Comment by AlkaMotors** [comment]
Replaced by pr#89
